### DISCUSSION TITLE
How to Retrieve Provider Access Token After Linking GitHub Account?

### DISCUSSION THREAD

**Discussion by a7medalyapany** [supabase/supabase#22686] [Questions]
I've got a setup where users initially log in using an email provider, which sets up a session. But here's the twist: I link their GitHub account later on, and I'm using the 'repo' scope when I do this. My question is, how can I grab the provider's access token after the GitHub account is linked? I need that token for further operations. Any insights or code snippets on how to pull this off would be super helpful! Thanks in advance!

**Comment by nishikaza** [comment]
@a7medalyapany did you get anywhere with this? Also working through a similar flow and looking for guidance here, thanks!

**Reply by a7medalyapany** [reply]
I already have a solution, but I believe there's a better one, of course. We could exchange the code that we getting back after linking identity in the auth/callback API and save it in the database, but this wouldn't be secure. Supabase advises against storing it in the database for security reasons, I guess. Alternatively, we could manually save it in the cookies, but these would expire, and you'd need to repeat the process, which obviously isn't efficient. So basically, it's a solution, but not the best one.

**Reply by nishikaza** [reply]
Are you doing the step of storing the tokens in the DB by just making changes in the auth/callback API file or are you making changes anywhere else in your code also? I'm leaning towards using cookies, and think if we were to save the provider refresh token, we could just make a call to Google's auth endpoint and receive a new access token.

**Reply by a7medalyapany** [reply]
So, like, in my situation I've made some tweaks in the auth/callback route yk? I did it based on this specific condition. rn i'm just saving it in the cookies, but it's not like permanent or anything. Also I tried to catch the provider refresh_token when using the linkIdentity function, but I couldn't grab it. However im getting the provider_token since I used exchangeCodeForSession passing the code that the providers give back.

```typescript
export async function GET(request: Request) {
  const requestUrl = new URL(request.url);
  const code = requestUrl.searchParams.get("code");
  const origin = requestUrl.origin;

  if (code) {
    const supabase = createClient();
    const { data, error } = await supabase.auth.exchangeCodeForSession(code);

    if (error) {
      console.error("Error exchanging code for session:", error.message);
      return NextResponse.error();
    }

    // Check if the session and token exist
    if (data.session && data.session.provider_token) {
      // You can log the session data to the console
      console.log("User:", data.user); //printed
      console.log("Session:", data.session.provider_token); //printed
      console.log("Refresh Token:", data.session.provider_refresh_token); //undefined

      const response = NextResponse.redirect(`${origin}/protected`);
      
      // Set a cookie with the provider access token
      const name = 'provider_token';
      const value = data.session.provider_token;
      console.log("Setting cookie:", name, value);


      response.cookies.set({ name, value});
      

      return response;
    }
  }

  // If there's no code or other issues, redirect to a fallback page
  return NextResponse.redirect(`${origin}/login`);
}
```

**Reply by nishikaza** [reply]
Helpful, thanks for the snippet! Isn't expected behavior that both provider_token and provider_refresh_token are returned though? Wonder what the issue is

**Reply by a7medalyapany** [reply]
I believe that the provider token is returned only once when you link the accounts. However, the provider_refresh_token is different because supabase deals with the main provider (in my case Email) as the primary entity and GitHub as the secondary. It just takes some information and saves it—nothing more. For now I guess we either put it in the database but that would affect performance since you have to make a request whenever you need it, or we put it in cookies. But then we’d have to make the user relink their account once the token expires, which isn't ideal either. Let me know if you have any ideas.

**Reply by nishikaza** [reply]
Right, my flow is OTP login via SMS and then linking a Google identity to the current user. I then want to use the Google provider_token to make future calls for actions on the Google API so it sounds like we have similar intended outcomes. Are you saying then that the provider_token getting registered is the phone provider, or in your case email? Or have you been able to access and use your secondary entity's provider token at all?

I saw this [page](https://supabase.com/docs/reference/javascript/auth-signinwithoauth?example=sign-in-with-scopes) referenced on the Supabase docs but am not totally clear on where to use the onAuthStateChange with our flow. Lmk if any thoughts

**Reply by a7medalyapany** [reply]
after linking the new provider I redirect the user to the `/auth/callback` and here I got the GitHub provider_token but only once immediately after the user linked their account, so atm we have to send them to our backend servers for further processing, 'cause we won't be able to access this again unless we redo this process again tho.
in your case whenever you press the relink button redirects the user to `/auth/callback` and try to run some code based on some conditions to not run it every time you redirect the user to `/auth/callback`, like check for the identities and if it includes google do those steps
otherwise go with the normal behavior. hope i made it clear

**Reply by nishikaza** [reply]
How are you doing any further processing if you don't have the provider refresh token though? Sorry if I'm not understanding but don't you need this to make any call's to the provider's API? At my current state, I'm not able to make any call's to the provider's API, but my flow is after the user has signed in with SMS OTP, I'm checking if the user's identities already includes Google and if not, I'm calling signInWithOAuth, but if it does not, then I'm calling linkIdentity to link the Google provider for the first time. 

Does this sound right to you? I appreciate your help here, and apologies if I'm overlooking something basic.

**Reply by a7medalyapany** [reply]
im clueless about how to obtain the refresh_token from the secondary provider. That's what im hoping to get advice on here. let me know if you've figured anything out yet.

**Reply by nishikaza** [reply]
Yeah, I got it working. The provider refresh token is only returned the very first time the user authenticates with your app. You can test this out by deleting a user in auth.users in the dashboard and linking the provider identity for the first time. You can then check if provider_refresh_token is being returned in `auth/callback/route.ts` and store this value if it is being returned. I had to also insert the offline `access_type` in options in my `SignInWithOAuthCredentials` parameter, which is passed to `LinkIdentity()` like this:   
 ```
  const param: SignInWithOAuthCredentials = {
    provider: provider as Provider,
    options: {
      redirectTo: `${window.location.origin}/auth/callback?next=/action`,
      scopes: YTScope,
      queryParams: {
        access_type: "offline",
      },
    },
  };

  const { data, error } = await supabase.auth.linkIdentity(param);
```
 Hope this helps!

**Reply by a7medalyapany** [reply]
I did a bit of searching, and I believe the tokens we receive don't expire unless the user revokes them. However, some providers don't follow the same rules. In my case, I'm referring to GitHub, so you should check your provider's policy.
